src/boxrobots/src/controllers/QPinnerloopcontroller.jl

```diff
@@ -0,0 +1,476 @@
+using PyCall
+@pyimport boxatlas
+@pyimport boxatlas.boxatlas as boxatlastypes
+
+abstract ContactSwitchType
+immutable NoSwitch <: ContactSwitchType end
+immutable MakingContact <: ContactSwitchType end
+immutable BreakingContact <: ContactSwitchType end
+
+type ContactSwitch{T<:ContactSwitchType}
+  limb_symbol::Symbol
+end
+
+function ContactSwitch{T}(limb_symbol::Symbol, ::Type{T})
+  return ContactSwitch{T}(limb_symbol)
+end
+
+function ContactSwitch(cs_0::ContactState, cs_1::ContactState)
+  """
+  Compute the contact switch type when transitioning from
+  cs_0 --> cs_1
+  """
+
+  num_contact_state_mismatches = 0
+  limb_symbol = Symbol()
+  contact_switch_type = ContactSwitchType
+
+  for (limb_sym, contact_indicator_0) in cs_0
+    contact_indicator_1 = cs_1[limb_sym]
+    # there was a contact switch
+    if contact_indicator_0 != contact_indicator_1
+      if (contact_indicator_0 & !contact_indicator_1)
+        contact_switch_type = BreakingContact
```

Comment:
since each contact switch type is a different Julia type, this will create a type-instability  https://docs.julialang.org/en/stable/manual/performance-tips/#write-type-stable-functions 

That's not a big problem, it will just make this code run more slowly than it otherwise would. If it makes your life easier to use different types and then let multiple dispatch make decisions for you, then that's fine. But if this ends up being a performance bottleneck, then you might want to consider using an `@enum` instead